src/config.rs

```diff
@@ -0,0 +1,215 @@
+use std::collections::BTreeMap;
+use std::path::{Path, PathBuf};
+use std::fs::File;
+use std::io::Read;
+use toml::{self, Value};
+use serde::Deserialize;
+
+use errors::*;
+
+
+/// The overall configuration object for MDBook.
+#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
+#[serde(default)]
+pub struct Config {
+    /// Metadata about the book.
+    pub book: BookConfig,
+    /// Arbitrary information which renderers can use during the rendering
+    /// stage.
+    pub output: BTreeMap<String, Value>,
+    /// Information for use by preprocessors.
+    pub preprocess: BTreeMap<String, Value>,
+    /// Information for use by postprocessors.
+    pub postprocess: BTreeMap<String, Value>,
+}
+
+impl Config {
+    /// Load a `Config` from some string.
+    pub fn from_str(src: &str) -> Result<Config> {
+        toml::from_str(src).chain_err(|| Error::from("Invalid configuration file"))
+    }
+
+    /// Load the configuration file from disk.
+    pub fn from_disk<P: AsRef<Path>>(config_file: P) -> Result<Config> {
+        let mut buffer = String::new();
+        File::open(config_file)
+            .chain_err(|| "Unable to open the configuration file")?
+            .read_to_string(&mut buffer)
+            .chain_err(|| "Couldn't read the file")?;
+
+        Config::from_str(&buffer)
+    }
+
+    /// Convenience method for getting the html renderer's configuration.
+    ///
+    /// # Note
+    ///
+    /// This is for compatibility only. It will be removed completely once the
+    /// rendering and plugin system is established.
+    pub fn html_config(&self) -> Option<HtmlConfig> {
+        self.try_get_output("html").ok()
+    }
+
+    /// Try to get an output and deserialize it as a `T`.
+    pub fn try_get_output<'de, T: Deserialize<'de>, S: AsRef<str>>(&self, name: S) -> Result<T> {
+        get_deserialized(name, &self.output)
+    }
+
+    /// Try to get the configuration for a preprocessor, deserializing it as a
+    /// `T`.
+    pub fn try_get_preprocessor<'de, T: Deserialize<'de>, S: AsRef<str>>(&self, name: S) -> Result<T> {
+        get_deserialized(name, &self.preprocess)
+    }
+
+    /// Try to get the configuration for a postprocessor, deserializing it as a
+    /// `T`.
+    pub fn try_get_postprocessor<'de, T: Deserialize<'de>, S: AsRef<str>>(&self, name: S) -> Result<T> {
+        get_deserialized(name, &self.postprocess)
+    }
+}
+
+/// Convenience function to load a value from some table then deserialize it.
+fn get_deserialized<'de, T: Deserialize<'de>, S: AsRef<str>>(name: S, table: &BTreeMap<String, Value>) -> Result<T> {
+    match table.get(name.as_ref()) {
+        Some(output) => output
+            .clone()
+            .try_into()
+            .chain_err(|| "Couldn't deserialize the value"),
+        None => bail!("Key Not Found"),
+    }
```

Comment:
Can these two match arms produce a more helpful error message?
It is only helpful for library users, but still.

Reply:
Yeah good point. Although the deserializing error is chained, meaning if you iterate through the causes you can see the root reason for why deserializing failed.  Giving you a high level "deserializing failed" message, plus a more low level error if you need more information.

I'm not sure what to do in the `None` case though. I really want to avoid returning an `Option<Result<T>>` if I can, that just feels ugly.